### PR TITLE
Add /filter test

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

File: server/chat-plugins/chat-monitor.ts [hunk 1]
```diff
@@ -667,8 +667,65 @@ export const commands: ChatCommands = {
...
+		this.sendReplyBox(
+			`<code>/test[monitorname]filter [test string]</code> OR <code>/testfilter [test string] | [monitorname]</code>:<br />Tests whether or not the provided test string would trigger the respective monitor.<br />` +
+			`All usable commands: <code>${monitorNames.map(x => `/test${x.replace('filter', '')}filter`).join('</code>, <code>')}</code><br />` +
+			`Can only be broadcast in Staff and Upper Staff. Requires: % @ &`
+		);
+	},
 };
 
+// Outside of `process.nextTick()` because it causes test to fail otherwise
+for (const list of Object.keys(Chat.monitors).filter(x => Chat.monitors[x].monitor)) {
```
Comment (mia-pi-git) [inline server/chat-plugins/chat-monitor.ts:725]: This needs to be in nextTick to ensure all monitors are loaded.

Comment (KrisXV) [reply server/chat-plugins/chat-monitor.ts:725]: it causes `npm test` to fail atrociously

```
 !

  0 passing (2ms)
  1 failing

  1) Uncaught error outside test suite:
     Uncaught TypeError: Cannot set property 'testautolockfilter' of undefined
      at C:\[redacted]\showdown\.server-dist\chat-plugins\chat-monitor.js:726:60
      at processTicksAndRejections (node:internal/process/task_queues:76:11)
```

File: server/chat-plugins/chat-monitor.ts [hunk 2]
```diff
@@ -667,8 +667,65 @@ export const commands: ChatCommands = {
 		}
 		this.globalModlog(`ALLOWNAME`, target);
 	},
+	testfilter(target, room, user, connection, cmd) {
+		this.privatelyCheckCan('lock');
+		this.checkBroadcast(true);
+		if (room && ['staff', 'upperstaff'].includes(room.roomid)) this.runBroadcast();
```
Comment (AnnikaCodes) [inline server/chat-plugins/chat-monitor.ts:673]: Why checkBroadcast and then runBroadcast if it's in Staff? Can't you just `if (room && ['staff', 'upperstaff'].includes(room.roomid)) this.runBroadcast(true);`?

File: server/chat-plugins/chat-monitor.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-plugins/chat-monitor.ts:725]: Attach it to Chat.commands instead? I would assume that would kill two birds with one stone (It's what I did locally when we discussed this).

File: server/chat-plugins/chat-monitor.ts [hunk 3]
```diff
@@ -667,8 +667,65 @@ export const commands: ChatCommands = {
...
+				monitorName = monitorName.replace('filter', '');
+			}
+		}
+		// namefilter doesn't have a monitor function
+		if (!(monitorName && message) || !Chat.monitors[monitorName]?.monitor) {
+			return this.parse(`${this.cmdToken}help testfilter`);
+		}
+		const monitor = Chat.monitors[monitorName].monitor!;
+		const lcMessage = Chat.stripFormatting(message
+			.replace(/\u039d/g, 'N').toLowerCase()
```
Comment (AnnikaCodes) [inline server/chat-plugins/chat-monitor.ts:687]:
```suggestion
			.replace(/\u039d/g, 'N')
			.toLowerCase()
```

Comment (AnnikaCodes) [reply server/chat-plugins/chat-monitor.ts:687]: TBH this should probably be its own function.

File: server/chat-plugins/chat-monitor.ts [hunk 4]
```diff
@@ -667,8 +667,65 @@ export const commands: ChatCommands = {
...
+			.replace(/\u0430/g, 'a')
+			.replace(/\u0435/g, 'e')
+			.replace(/\u039d/g, 'e'));
+		let htmlBoxMessage = ``;
+		for (const line of Chat.filterWords[monitorName]) {
+			const ret = monitor.call(this, line, room, user, message, lcMessage, true);
+			if (typeof ret === 'string') {
+				htmlBoxMessage = ret;
+				break;
+			}
```
Comment (AnnikaCodes) [inline server/chat-plugins/chat-monitor.ts:701]:
```suggestion
			if (typeof ret === 'string') {
				htmlBoxMessage = ret;
				break;
			} else if (ret === false) {
			   htmlBoxMessage = `"${message}" would be blocked from being sent.`;
			   break;
			}
```

**Comment by AnnikaCodes** [comment]
Does it work in `process.nextTick` if you modify `Chat.commands` instead of the scoped `commands` object?

**Comment by mia-pi-git** [comment]
Honestly, why not put it under the `/filter` namespace? Then you could more easily add specifying the filter in the command (ie `/filter testpublicwarn`)

File: server/chat-plugins/chat-monitor.ts (hunk 2 above)
Comment (KrisXV) [reply server/chat-plugins/chat-monitor.ts:673]: bc i am dumb and forgot to remove the line

File: server/chat-plugins/chat-monitor.ts (hunk 3 above)
Comment (KrisXV) [reply server/chat-plugins/chat-monitor.ts]: agree

**Review by AnnikaCodes — APPROVED** [review]
This looks fine to me.

File: server/chat-plugins/chat-monitor.ts [hunk 5]
```diff
@@ -672,4 +733,8 @@ export const commands: ChatCommands = {
 process.nextTick(() => {
 	Chat.multiLinePattern.register('/filter (add|remove) ');
 	loadFilters();
+	for (const list of Object.keys(Chat.monitors).filter(x => Chat.monitors[x].monitor)) {
+		(Chat.commands['filter'] as ChatCommands)[`test${list.replace('filter', '')}`] = 'test';
+		if (list.includes('filter')) (Chat.commands['filter'] as ChatCommands)[`test${list}`] = 'test';
+	}
```
Comment (Zarel) [inline server/chat-plugins/chat-monitor.ts:739]: You could have the syntax be `/filter test [filtername]` to get around the plugin loading order issue.

File: server/chat-plugins/chat-monitor.ts [hunk 6]
```diff
@@ -642,13 +642,74 @@ export const commands: ChatCommands = {
...
+			const lcMessage = Chat.stripFormatting(message
+				.replace(/\u039d/g, 'N')
+				.toLowerCase()
+				// eslint-disable-next-line no-misleading-character-class
+				.replace(/[\u200b\u007F\u00AD\uDB40\uDC00\uDC21]/g, '')
+				.replace(/\u03bf/g, 'o')
+				.replace(/\u043e/g, 'o')
+				.replace(/\u0430/g, 'a')
+				.replace(/\u0435/g, 'e')
+				.replace(/\u039d/g, 'e'));
```
Comment (AnnikaCodes) [inline server/chat-plugins/chat-monitor.ts:671]: I take it you're not planning to break this out into its own function with this PR? I can do that myself then.

Comment (KrisXV) [reply server/chat-plugins/chat-monitor.ts:671]: Yeah it's not really relevant to this PR so I didn't feel too comfortable compacting it

File: server/chat-plugins/chat-monitor.ts (hunk 5 above)
Comment (KrisXV) [reply server/chat-plugins/chat-monitor.ts]: did @Zarel

File: test/server/chat-plugins/chat-monitor.js [hunk 7]
```diff
@@ -57,7 +57,6 @@ describe('Chat monitor', () => {
 			this.user.joinRoom(this.room.roomid, this.connection);
 
 			this.parse = async function (message) {
-				Chat.loadPlugins();
```
Comment (mia-pi-git) [inline test/server/chat-plugins/chat-monitor.js:60]: Maybe restore this?

Comment (KrisXV) [reply test/server/chat-plugins/chat-monitor.js]: I already did locally, not my fault my wifi sucks!!!

File: server/chat-plugins/chat-monitor.ts [hunk 8]
```diff
@@ -642,13 +642,71 @@ export const commands: ChatCommands = {
 			this.parse(`/join view-filters`);
 		},
 		help(target, room, user) {
-			this.parse(`/help filter`);
+			this.run(`filterhelp`);
+		},
+		test(target, room, user) {
+			this.privatelyCheckCan('lock');
```
Comment (Zarel) [inline server/chat-plugins/chat-monitor.ts:648]:
```suggestion
			this.checkCan('lock');
```

No longer needs to be private now that the individual subcommands aren't aliases.

File: server/chat-plugins/chat-monitor.ts [hunk 9]
```diff
@@ -642,13 +642,71 @@ export const commands: ChatCommands = {
...
+			if (htmlBoxMessage) {
+				return this.sendReplyBox(Chat.formatText(htmlBoxMessage, false, true));
+			} else {
+				throw new Chat.ErrorMessage(
+					`"${message}" doesn't trigger any filters on the ${monitorName}${monitorName.endsWith('filter') ? '' : ' filter'}. Check spelling?`
+				);
+			}
+		},
+		testhelp(target, room, user) {
+			this.privatelyCheckCan('lock');
```
Comment (Zarel) [inline server/chat-plugins/chat-monitor.ts:689]:
```suggestion
			this.checkCan('lock');
```

**Review by Zarel — APPROVED** [review]
looks good other than those two nits

**Comment by KrisXV** [comment]
Done @Zarel

**Comment by Zarel** [comment]
tyty

**Pull request merged**